modules/tinymce/src/core/main/ts/newline/InsertBlock.ts

```diff
@@ -242,139 +224,198 @@ const addBrToBlockIfNeeded = function (dom, block) {
-          } else {
-            caretNode = clonedNode;
-            block.appendChild(clonedNode);
-          }
+// Used as an internal function below. Dragging it up here to make the function state explicit (and read-only).
+interface CreateNewBlock {
+  container: Node;
+  schema: Schema;
+  parentBlockName: string;
+  newBlockName: string;
+  dom: DOMUtils;
+  parentBlock: Element;
+  editor: Editor;
+  editableRoot: Node;
+}
+
+const _createNewBlock = ({ container, schema, parentBlockName, newBlockName, dom, parentBlock, editor, editableRoot }: CreateNewBlock, name: string | undefined): Node => {
+  let node = container;
+  const textInlineElements = schema.getTextInlineElements();
+
+  const block = (name || parentBlockName === 'TABLE' || parentBlockName === 'HR') ? dom.create(name || newBlockName) : parentBlock.cloneNode(false);
+
+  let caretNode = block;
+
+  if (Settings.shouldKeepStyles(editor) === false) {
+    dom.setAttrib(block, 'style', null); // wipe out any styles that came over with the block
+    dom.setAttrib(block, 'class', null);
+  } else {
+    // Clone any parent styles
+    do {
+      if (textInlineElements[node.nodeName]) {
+        // Ignore caret or bookmark nodes when cloning
+        if (isCaretNode(node) || Bookmarks.isBookmarkNode(node)) {
+          continue;
         }
-      } while ((node = node.parentNode) && node !== editableRoot);
-    }
 
-    setForcedBlockAttrs(editor, block);
+        const clonedNode = node.cloneNode(false);
+        dom.setAttrib(clonedNode, 'id', ''); // Remove ID since it needs to be document unique
 
-    emptyBlock(caretNode);
+        if (block.hasChildNodes()) {
+          clonedNode.appendChild(block.firstChild);
+          block.appendChild(clonedNode);
+        } else {
+          caretNode = clonedNode;
+          block.appendChild(clonedNode);
+        }
+      }
+    } while ((node = node.parentNode) && node !== editableRoot);
+  }
 
-    return block;
-  };
+  setForcedBlockAttrs(editor, block);
 
-  // Returns true/false if the caret is at the start/end of the parent block element
-  const isCaretAtStartOrEndOfBlock = function (start?) {
-    let node, name;
+  emptyBlock(caretNode);
 
-    const normalizedOffset = normalizeZwspOffset(start, container, offset);
+  return block;
+};
 
-    // Caret is in the middle of a text node like "a|b"
-    if (NodeType.isText(container) && (start ? normalizedOffset > 0 : normalizedOffset < container.nodeValue.length)) {
-      return false;
-    }
+// Used as an internal function below. Dragging it up here to make the function state explicit (and read-only).
+interface IsCaretAtStartOrEndOFBlock {
+  offset: number;
+  container: Node;
+  isAfterLastNodeInContainer: boolean;
+  parentBlock: Element;
+  nonEmptyElementsMap: SchemaMap;
+}
+
+const _isCaretAtStartOrEndOfBlock = ({ offset, container, isAfterLastNodeInContainer, parentBlock, nonEmptyElementsMap }: IsCaretAtStartOrEndOFBlock, start: boolean): boolean => {
```

Comment:
See my points in the previous comment about prefixing with an underscore and destructing all arguments in a function argument list.